include/sparrow/mp_utils.hpp

```diff
@@ -14,19 +14,228 @@
+    using constify_t = typename constify<T, is_const>::type;
+
+    //////////////////////////////////////////////////
+    //// Type-predicates /////////////////////////////
+
+    /// Matches template types which can be used as type-wrappers for evaluation in type-predicates.
+    template< template< class > class W, class T >
+    concept type_wrapper = std::same_as<W<T>, typelist<T>>
+                        or std::same_as<W<T>, std::type_identity_t<T>>
+                        ;
+
+    /// Matches template types that can be evaluated at compile-time similarly to `std::true/false_type`
+    /// This makes possible to use predicates provided by the standard.
+    template< template<class> class P, class T >
+    concept ct_type_predicate = requires
+    {
+        { P<T>::value } -> std::convertible_to<bool>;
+    };
+
+    /// Matches types which instance can be called with an object representing a type to evaluate it.
+    /// The object passed to calling the predicate instance is either a typelist with one element or
+    /// `std::type_identity_t<T>`. Basically a value wrapper for representing a type.
+    /// This is useful to detect type predicates which allow being called like normal functions.
+    template< class P, class T >
+    concept callable_type_predicate = std::semiregular<P>
+        and (
+                requires(P predicate)
+                {
+                    { predicate(typelist<T>{}) } -> std::convertible_to<bool>;
+                }
+            or
+                requires(P predicate)
+                {
+                    { predicate(std::type_identity_t<T>{}) } -> std::convertible_to<bool>;
+                }
+            )
+        ;
+
+
+
+    /// Evaluates the provided compile-time template class predicate P given a type T, if P is of a similar shape to `std::true/false_type`.
+    template<class T, template<class> class P>
+        requires ct_type_predicate<P,T>
+    consteval
+    bool evaluate(P<T>)
+    {
+        return P<T>::value;
+    }
+
+    /// Evaluates the provided compile-time template class predicate P given a type T, if P's instance is callable given a value type wrapper of T.
+    template<class T, callable_type_predicate<T> P>
+    consteval
+    bool evaluate(P predicate)
+    {
+        if constexpr (requires (P p){ { p(typelist<T>{}) } -> std::same_as<bool>; })
+        {
+            return predicate(typelist<T>{});
+        }
+        else
+        {
+            return predicate(std::type_identity_t<T>{});
+        }
+    }
+
+    namespace predicate
+    {
+
+        /// Compile-time type predicate: `true` if the evaluated type is the same as `T`.
+        /// Example:
+        ///     static constexpr auto some_types = typelist<int, float>{}
+        ///     static constexpr auto same_as_int = same_as<int>{};
+        ///     static_assert( any_of(some_types, same_as_int) == true);
+        ///     static_assert( all_of(some_types, same_as_int) == false);
```

Comment:
I guess we can add this snippet to the tests.

Reply:
The tests currently do something similar but not with the named constexpr object. You would prefer it to be state the same?

Reply:
OK I admittedly did not check the tests when I wrote this comment. To me, the definition of `any_of` and or `all_of` are sufficiently clear without this comment but we can keep it.

Reply:
No problem, I added the same code in the checks still as it doesnt cost anything and shows it can be done with constexpr objects too.